pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/BaseIndexHandler.java

```diff
@@ -93,6 +100,24 @@ protected void createForwardIndexIfNeeded(SegmentDirectory.Writer segmentWriter,
       _tmpForwardIndexColumns.add(columnName);
     }
 
+    // Update the segmentMetadata
+    File indexDir = _segmentMetadata.getIndexDir();
```

Comment:
As per my understanding, whenever we update metadata properties file, we also want to update the in-memory segmentMetadata data-structure. Is that correct?

Can we move the reloadSegmentMetadata() call closer to the place where the we actually update the file - or maybe wrap in-memory and file updates into a common function that can be called?

This will also make sure that future callers don't miss updating in-memory DS when updating the file.

Reply:
> As per my understanding, whenever we update metadata properties file, we also want to update the in-memory segmentMetadata data-structure. Is that correct?

Yes that's my understanding as well. 

> 

So one problem is that in the `BaseIndexHandler` I have the following code (unimportant code removed):

```
    invertedIndexAndDictionaryBasedForwardIndexCreator.regenerateForwardIndex();

...
...

    // Update the segmentMetadata
    File indexDir = _segmentMetadata.getIndexDir();
    reloadSegmentMetadata(indexDir, columnName);
```

`invertedIndexAndDictionaryBasedForwardIndexCreator.regenerateForwardIndex()` calls the metadata file update part. Even if I moved the in-memory update part closer to the file update in this scenario, I'll need to call `_segmentMetadata = new SegmentMetadataImpl(indexDir);` in this function since we're creating a new `SegmentMetadata` object, similar to the reason why I need to do this in `SegmentPreProcessor`. Is that okay?

Reply:
I was thinking we could just change the signature of `updateMetadataProperties` as follows:

`public static SegmentMetadata  updateMetadataProperties(SegmentMetadata segmentMetadata, Map<String, String> metadataProperties, File indexDir)`

Now we can just call,
`_segmentMetadata = updateMetadataProperties(...) `everywhere. 

Does that work?

Reply:
That won't solve the issue where `_segmentMetadata` assigned inside the class won't reflect the new object outside the class :( 

The problem is even if I make your suggested change, I'll have to do the following:

```
  public static SegmentMetadata updateMetadataProperties(SegmentMetadata segmentMetadata, Map<String,
      String> metadataProperties, File indexDir, SegmentDirectory segmentDirectory)
      throws IOException {
    PropertiesConfiguration propertiesConfiguration = SegmentMetadataUtils.getPropertiesConfiguration(segmentMetadata);
    for (Map.Entry<String, String> entry : metadataProperties.entrySet()) {
      propertiesConfiguration.setProperty(entry.getKey(), entry.getValue());
    }
    SegmentMetadataUtils.savePropertiesConfiguration(propertiesConfiguration);

    segmentMetadata = new SegmentMetadataImpl(indexDir);
    segmentDirectory.reloadMetadata();

    return segmentMetadata;
  }
```

Since a new object is created and returned and we assign that like this: `_segmentMetadata = updateMetadataProperties(...) `, it's still replacing the local class object and not the object outside the class.

Reply:
Yeah, checked your latest commit and it looks better now. Agreed that we'd still have to update the object outside the class to reflect the new metadata. Thanks for the making the change.